### PR TITLE
feat: add callback for session timeout

### PR DESCRIPTION
**Pull request by mattmazzola**
Fixes ADO#1181

The task said to replace the number with function, but that would mean calling the function every input to see if session should be timed out which seemed very similar to someone creating an end session action and learning to invoke it when entity is set.

I think this combo is still setting inactive time + function allows the function to only be called when the timeout expires.

I'm not sure if this solves the case you wanted it to.
Basically mimics the way we register entity detection callback.

### CODE REVIEW

File: src/CLRunner.ts [hunk 1]
```diff
@@ -433,7 +439,8 @@ export class CLRunner {
                 const currentTicks = new Date().getTime();
                 let lastActive = await state.BotState.GetLastActive()
                 let passedTicks = currentTicks - lastActive;
-                if (passedTicks > this.modelOptions.sessionTimout!) {
+                if (passedTicks > this.modelOptions.sessionTimout) {
+                    await this.CallSessionTimeoutCallback(state, [])
```
Comment (mattmazzola) [inline src/CLRunner.ts]: This is currently the problem. We need to pass current list of entities to create the memory manager.

I saw in the EndSessionCallback there was extra request to the server to query the entities from the model, but that gets the entities from latest package which may be out of sync with the running bot.

We could do this something similar to this but it seems like it sets up bad practice of the SDK calling the server. Given this is already the timeout and we don't care about latency as much, wanted to make sure it's the only option.

https://github.com/microsoft/ConversationLearner-SDK/blob/master/src/CLRunner.ts#L374

**Comment by mattmazzola** [comment]
Seems there needs to be more thought about use cases. Lars implied the EndSessionCallback has overlapping function. (I think, if the session expires due to timeout, session will end, calling the end session callback)

Questions:
When are these functions invoked, or under what different conditions?

A use case I can see is taking different behavior on session expiration than a normal end session.
For example when session expires you might want to save off conversation state or other data so you can restore it, user might have left unexpectedly in middle of bot interaction. However on normal end session you just want to reset the memory.

Instead of having two different callbacks, I think it could be better to have a single (existing) end session callback and passing in the reason it was called. Reasons we have so far are "expiration" or "action".
If it's action, they can just the data string for more detailed info.
